### ISSUE TITLE
Use pinpoint about aliyun message service

### ISSUE THREAD

**Issue by linux0x5c** [naver/pinpoint#4882]
I have a pinpoint question I need to ask
you.When I added the pinpoint agent to tomcat，I use aliyun sdk to call
aliyun message service.Unfortunately，a socket timeout occurred.
Below is my system environment:
system version:  CentOS Linux release 7.4.1708 (Core)
pinpoint agent version： 1.8.0
jdk version: java version "1.7.0_79"

My java method:

    public static void testPinPoint(String postUrl) {
    	CloseableHttpAsyncClient httpclient = HttpAsyncClients.custom().build();
    	httpclient.start();
    	HttpPost httpPost = new HttpPost(postUrl);
    	
    	//custom repeat stream entity
    	BasicHttpEntity basicHttpEntity = new BasicHttpEntity(){
    		@Override
    		public boolean isRepeatable() {
    			// TODO Auto-generated method stub
    			return this.getContent().markSupported();
    		}
    		
    		@Override
    	    public void writeTo(final OutputStream outstream) throws IOException {
    	        Args.notNull(outstream, "Output stream");
    	        System.out.println("writeTo::::::::::::::::::::::::::::");
    	        final InputStream instream = this.getContent();
    	        instream.reset();
    	        try {
    	            final byte[] buffer = new byte[OUTPUT_BUFFER_SIZE];
    	            int l;
    	            if (this.getContentLength() < 0) {
    	                // consume until EOF
    	                while ((l = instream.read(buffer)) != -1) {
    	                    outstream.write(buffer, 0, l);
    	                }
    	            } else {
    	                // consume no more than length
    	                long remaining = this.getContentLength();
    	                while (remaining > 0) {
    	                    l = instream.read(buffer, 0, (int)Math.min(OUTPUT_BUFFER_SIZE, remaining));
    	                    if (l == -1) {
    	                        break;
    	                    }
    	                    outstream.write(buffer, 0, l);
    	                    remaining -= l;
    	                }
    	            }
    	        } finally {
    	            //instream.close();
    	        }
    	    }
    	};
    	
    	byte[] bytearray = "123456".getBytes();
    	basicHttpEntity.setContent(new ByteArrayInputStream("123456".getBytes()));
    	basicHttpEntity.setContentLength(bytearray.length);
    	//when set contentType socket will be timeout
    	basicHttpEntity.setContentType("text/plain;charset=UTF-8");
    	basicHttpEntity.setChunked(false);
    	
    	httpPost.setEntity(basicHttpEntity);
    	Future<HttpResponse> future = httpclient.execute(httpPost, new FutureCallback<HttpResponse>(){

			@Override
			public void completed(HttpResponse result) {
			}
			@Override
			public void failed(Exception ex) {
				ex.printStackTrace();
			}

			@Override
			public void cancelled() {
			}
    		
    	});
    	try{
    		future.get();
    	}catch (Exception e) {
			// TODO: handle exception
    		e.printStackTrace();
		}
    	System.out.println("sucess.............");
    	
    }

When I set the http header ContentType with aliyun SDK,the request
always is timeout.When I remove the pinpoint agent in tomcat, the
socket will not time out.The strange thing is that I added the
pinpoint agent to tomcat and did not set the HTTP header ContentType
using the aliyun SDK, and the socket timeout did not occur.I don't
know where the problem is, I hope you can help me.

My pinpoint configuration tomcat log and pinpoint agent log are placed
in the attachment.
[logs.zip](https://github.com/naver/pinpoint/files/2574875/logs.zip)

**Comment by Xylus** [comment]
Hi @linux0x5c 
I'm not sure if this is related but may be aliyun does not accept custom HTTP headers? (Pinpoint adds custom HTTP headers to propagate trace)
Not sure if this is the correct documentation for your use case but it looks like you need to open a ticket to allow custom headers - https://www.alibabacloud.com/help/doc-detail/65094.htm

You can check Pinpoint's custom HTTP headers [here](https://github.com/naver/pinpoint/blob/master/bootstrap-core/src/main/java/com/navercorp/pinpoint/bootstrap/context/Header.java).

Still not entirely sure why it doesn't timeout if *ContentType* isn't set, but maybe you could try it out?

**Comment by linux0x5c** [comment]
Hi @Xylus 

I re-tested, this time I did not use Alibaba Cloud's messaging service, but used the above code to simulate requesting Alibaba Cloud messaging service. But the problem still exists, as long as I add http header ContentType to the constructed request message. My message still can't be sent out, and socket timeout still occurs. This shows that there is no relationship with Alibaba Cloud's message service verification. I suspect that pinpoint does not support this streaming data request tracking.

lab environment:
Tomcat version: apache-tomcat-7.0.63
Client libary: Apache http client 4.x
Code: The code above this issue

**Comment by Xylus** [comment]
@linux0x5c 
Ah gotcha, so it seems it's not related to Alibaba Cloud's messaging service, right?
I'll take a look into this and see what's going on.
Thanks!

**Comment by linux0x5c** [comment]
@Xylus 
Yes, it has nothing to do with Alibaba Cloud's messaging service.
Looking forward to your good news.
Thanks!

**Comment by Xylus** [comment]
Hi @linux0x5c 
Here's what I found.

### What is happening?
The server is returning 500 due to *Timeout attempting to read data from the socket*.
The reason for this is because pinpoint agent is consuming the content in **HttpClient4EntityExtractor** (see below) so the client is not writing the content out the socket.
```
    private static String entityUtilsToString(final HttpEntity entity, final String defaultCharset, final int maxLength) throws Exception {
        ...
        final FixedByteArrayOutputStream outStream = new FixedByteArrayOutputStream(maxLength);
        entity.writeTo(outStream);
        final String entityValue = outStream.toString(charset);
        ...
        return entityValue;
    }
```
However, this is only done if the entity is **repeatable**.
```
    @Override
    public String getEntity(HttpRequest httpRequest) {
        ...
        if (entity != null && entity.isRepeatable() && entity.getContentLength() > 0) {
            return entityUtilsToString(entity, Charsets.UTF_8_NAME, 1024);
        }
        ...
    }
```
Consuming a none repeatable entity's content (via `writeTo(...)` in this case) will result in the client not writing the content out the socket; and the server, which is expecting data to arrive, will throw a timeout since nothing is received.

### But the entity in the code above is repeatable
Your **BasicHttpEntity** above is repeatable as can be seen below.
```
    ...
    //custom repeat stream entity
    BasicHttpEntity basicHttpEntity = new BasicHttpEntity() {
        @Override
        public boolean isRepeatable() {
            // TODO Auto-generated method stub
                return this.getContent().markSupported();       // ByteArrayInputStream.markSupported() returns true
        }
        ...
    }
    ...
```
However, there is another contract specified in **HttpEntity**'s `getContent()` that your `basicHttpEntity` is missing.
```
   /**
     * Returns a content stream of the entity.
     * {@link #isRepeatable Repeatable} entities are expected
     * to create a new instance of {@link InputStream} for each invocation
     * of this method and therefore can be consumed multiple times.
       ...
     */
    InputStream getContent() throws IOException, UnsupportedOperationException;
```
The default implementation of **BasicHttpEntity**'s `getContent()` method is to simply return the underlying **InputStream**.
Your *repeatable* `basicHttpEntity` does not override the `getContent()` method, and so is returning the underlying **InputStream** directly.
In order to make your `basicHttpEntity` repeatable, you will have to override the `getContent()` method as well, and have it create a new instance of the underlying **InputStream** and return that instead.

Something like below should solve your issue. (Example code, please modify it to suit your needs)
```
    ...
    //custom repeat stream entity
    BasicHttpEntity basicHttpEntity = new BasicHttpEntity() {
        ...
        @Override
        public InputStream getContent() throws IllegalStateException {
            InputStream is = super.getContent();
            if (is.markSupported()) {
                try {
                    ByteArrayOutputStream baos = new ByteArrayOutputStream();
                    byte[] buffer = new byte[1024];
                    int len;
                    while ((len = is.read(buffer)) > -1) {
                        baos.write(buffer, 0, len);
                    }
                    baos.flush();
                    return new ByteArrayInputStream(baos.toByteArray());
                } finally {
                    is.reset();
                }
            } else {
                return is;
            }
        }
    };
```


#### PS. Why is it only happening when content type is set?
It's only happening when the content type is set because the **HttpClient4EntityExtractor**'s `entityUtilsToString(...)` method throws a **NullPointerException** and does not consume the content if the content type is not set.
```
    private static String entityUtilsToString(final HttpEntity entity, final String defaultCharset, final int maxLength) throws Exception {
        ...
        if (entity.getContentType().getValue().startsWith("multipart/form-data")) {            // NullPointerException here due to entity.getContentType() being null.
            return "content type is multipart/form-data. content length:" + entity.getContentLength();
        }
        ...
        final FixedByteArrayOutputStream outStream = new FixedByteArrayOutputStream(maxLength);
        entity.writeTo(outStream);            // This is where the content is consumed. Due to the NPE above, the content is not consumed if entity's content type is not set (is null).
        final String entityValue = outStream.toString(charset);
        ...
    }
```
We should add a null check in our **HttpClient4EntityExtractor** for this.
Note that this won't fix your issue - you will need to override `getContent()` method of your `basicHttpEntity` to get everything working.

Let me know if you have any questions.
Thanks.

**Comment by Xylus** [comment]
tldr; version
1. You should override `getContent()` and have it return a new instance of **InputStream**.
2. There needs to be a null check in **HttpClient4EntityExtractor** to check for entity's content type not being set (but this isn't the root cause of your issue).

**Comment by linux0x5c** [comment]
Hi @Xylus 

Thank you very much for your doubts. This is really very helpful for me. From your solution point of view, Alibaba's JAVA SDK does not fully consider this situation, we have already issued a work order to Alibaba Cloud, and they have been recognized. This problem has plagued me for a long time, and finally solved it today. Thank you very much again.
